### PR TITLE
Integrates soft-deletes into Elasticsearch

### LINKED ISSUE #29530: Use soft-deletes to maintain document history
With the introduction of [soft-delete]((https://issues.apache.org/jira/browse/LUCENE-8200)) in Lucene, a history of a document can be maintained. This meta issue tracks work on migrating from hard-deletes to soft-deletes.

Cut over
--------
- [x] Replace hard-updates by soft-updates (https://github.com/elastic/elasticsearch/pull/29458)
- [x] Replace hard-deletes by soft-deletes (https://github.com/elastic/elasticsearch/pull/29549)
- [x] Index no-op (https://github.com/elastic/elasticsearch/pull/30226) 
- [x] Index stale operations (i.e. out of order) on replicas (https://github.com/elastic/elasticsearch/pull/29679)

Retention and source
--------------------
- [x] Use SoftDeletesRetentionMergePolicy to maintain only the recent history (https://github.com/elastic/elasticsearch/pull/30335)
- [x] Special merge strategy when `_source` is disabled (https://github.com/elastic/elasticsearch/pull/31106)
- ~~Do not keep delete tombstones in version map. Delete tombstones can be queried from Lucene~~ (to be done in https://github.com/elastic/elasticsearch/issues/31637)

Translog
--------
- [x] Add API allows querying operations given a range of seq# (https://github.com/elastic/elasticsearch/pull/30120)
- ~~Use `maxTimestamp` from engine for `autoGeneratedIdTimestamp`~~ (migrated to #30086)
- [x] Replace translog by Lucene history in peer-recovery (https://github.com/elastic/elasticsearch/pull/30522)
- [x] Replace translog by Lucene history in primary-replica resync (#33178)
- [x] Only sends operations after the local checkpoint in peer-recovery (#33190)

TBD
-------
- ~~Avoid having multiple docs for the same stale operation. We currently defer the dedup until the search time but we might revisit this decision to do it at the index time~~(to be done in #31637). 

Misc
-------
- [x] Store the reason of noop in its document tombstone (https://github.com/elastic/elasticsearch/pull/30570)
- [x] Use exact numDocs in synced-flush (https://github.com/elastic/elasticsearch/pull/30228)
- [x] Randomize soft-deletes settings (https://github.com/elastic/elasticsearch/pull/31585)
- [x] Do not expose hard-deleted documents (#32333)

### PR DESCRIPTION
**Pull request by dnhatn**
This PR integrates Lucene soft-deletes[(LUCENE-8200)](https://issues.apache.org/jira/browse/LUCENE-8200) into Elasticsearch.
Highlight works in this PR include:

1. Replace hard-deletes by soft-deletes in InternalEngine
2. Use _recovery_source if _source is disabled or modified (#31106)
3. Soft-deletes retention policy based on the global checkpoint (#30335)
4. Read operation history from Lucene instead of translog (#30120)
5. Use Lucene history in peer-recovery (#30522)

These pieces were reviewed already in the feature branch but we 
would like to give them an extra look before pulling into the upstream.

Relates #30086
Closes #29530

These works have been done by the whole team; however, these individuals
(lexical order) have significant contribution in coding and reviewing:

Co-authored-by: Adrien Grand <jpountz@gmail.com>
Co-authored-by: Boaz Leskes <b.leskes@gmail.com>
Co-authored-by: Jason Tedor <jason@tedor.me>
Co-authored-by: Martijn van Groningen <martijn.v.groningen@gmail.com>
Co-authored-by: Nhat Nguyen <nhat.nguyen@elastic.co>
Co-authored-by: Simon Willnauer <simonw@apache.org>

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -2311,4 +2488,59 @@ public boolean isRecovering() {
...
+            this.refreshedCheckpoint = new AtomicLong(initialLocalCheckpoint);
+        }
+        @Override
+        public void beforeRefresh() {
+            pendingCheckpoint = localCheckpointTracker.getCheckpoint(); // All change until this point should be visible after refresh
+        }
+        @Override
+        public void afterRefresh(boolean didRefresh) {
+            if (didRefresh) {
+                refreshedCheckpoint.set(pendingCheckpoint);
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: can we add an assertion that we actually set the pendingCheckpoint?

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 2]
```diff
@@ -2311,4 +2488,59 @@ public boolean isRecovering() {
...
+            return super.softUpdateDocuments(term, docs, softDeletes);
+        }
+    }
+
+    /**
+     * Returned the last local checkpoint value has been refreshed internally.
+     */
+    final long lastRefreshedCheckpoint() {
+        return lastRefreshedCheckpointListener.refreshedCheckpoint.get();
+    }
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: extra newline please

**Review by s1monw — CHANGES_REQUESTED** [review]
Looks awesome. Great piece of work! I left some comments most of them are nits but a couple of minors.

File: server/src/main/java/org/elasticsearch/common/lucene/Lucene.java [hunk 3]
```diff
@@ -829,4 +835,82 @@ public int length() {
...
+                @Override
+                public LeafReader wrap(LeafReader leaf) {
+                    SegmentReader segmentReader = segmentReader(leaf);
+                    Bits hardLiveDocs = segmentReader.getHardLiveDocs();
+                    if (hardLiveDocs == null) {
+                        return new LeafReaderWithLiveDocs(leaf, null, leaf.maxDoc());
+                    }
+                    // TODO: Avoid recalculate numDocs everytime.
+                    int numDocs = 0;
+                    for (int i = 0; i < hardLiveDocs.length(); i++) {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/common/lucene/Lucene.java]: Yes, but there is something not clear on my side. I will add a comment then make the change in a follow-up so we can clarify things.

**Review by jpountz — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/index/IndexSettings.java [hunk 4]
```diff
@@ -237,6 +237,21 @@
...
+     */
+    public static final Setting<Boolean> INDEX_SOFT_DELETES_SETTING =
+        Setting.boolSetting("index.soft_deletes.enabled", false, Property.IndexScope, Property.Final);
+
+    /**
+     * Controls how many soft-deleted documents will be kept around before being merged away. Keeping more deleted
+     * documents increases the chance of operation-based recoveries and allows querying a longer history of documents.
+     * If soft-deletes is enabled, an engine by default will retain all operations up to the global checkpoint.
+     **/
+    public static final Setting<Long> INDEX_SOFT_DELETES_RETENTION_OPERATIONS_SETTING =
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/IndexSettings.java]: +1

File: server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java [hunk 5]
```diff
@@ -101,6 +104,9 @@ private void updateTranslogDeletionPolicy() throws IOException {
         assert minRequiredGen <= lastGen : "minRequiredGen must not be greater than lastGen";
         translogDeletionPolicy.setTranslogGenerationOfLastCommit(lastGen);
         translogDeletionPolicy.setMinTranslogGenerationForRecovery(minRequiredGen);
+
+        softDeletesPolicy.setLocalCheckpointOfSafeCommit(
+            Long.parseLong(safeCommit.getUserData().get(SequenceNumbers.LOCAL_CHECKPOINT_KEY)));
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java]: We ensure that all index commits should have this key.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 6]
```diff
@@ -2296,6 +2419,60 @@ long getNumDocUpdates() {
...
+     */
+    final long getMinRetainedSeqNo() {
+        assert softDeleteEnabled : Thread.currentThread().getName();
+        return softDeletesPolicy.getMinRetainedSeqNo();
+    }
+
+    @Override
+    public Closeable acquireRetentionLockForPeerRecovery() {
+        final Closeable translogLock = translog.acquireRetentionLock();
+        final Releasable softDeletesLock = softDeletesPolicy.acquireRetentionLock();
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Great catch. We should acquire one of them but not both.

File: server/src/main/java/org/elasticsearch/index/engine/RecoverySourcePruneMergePolicy.java [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+        throws IOException {
+        NumericDocValues recoverySource = reader.getNumericDocValues(recoverySourceField);
+        if (recoverySource == null || recoverySource.nextDoc() == DocIdSetIterator.NO_MORE_DOCS) {
+            return reader; // early terminate - nothing to do here since non of the docs has a recovery source anymore.
+        }
+        BooleanQuery.Builder builder = new BooleanQuery.Builder();
+        builder.add(new DocValuesFieldExistsQuery(recoverySourceField), BooleanClause.Occur.FILTER);
+        builder.add(retainSourceQuerySupplier.get(), BooleanClause.Occur.FILTER);
+        IndexSearcher s = new IndexSearcher(reader);
+        s.setQueryCache(null);
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/RecoverySourcePruneMergePolicy.java]: Done

File: server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+                parallelArray.isTombStone[index] = combinedDocValues.isTombstone(segmentDocID);
+                parallelArray.hasRecoverySource[index] = combinedDocValues.hasRecoverySource(segmentDocID);
+            }
+            // now sort back based on the shardIndex. we use this to store the previous index
+            ArrayUtil.introSort(scoreDocs, Comparator.comparingInt(i -> i.shardIndex));
+        }
+    }
+
+    private TopDocs searchOperations(ScoreDoc after) throws IOException {
+        final Query rangeQuery = LongPoint.newRangeQuery(SeqNoFieldMapper.NAME, fromSeqNo, toSeqNo);
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java]: Yes, I use "lastSeenSeqNo + 1" as the lower bound.

File: server/src/main/java/org/elasticsearch/index/IndexSettings.java [hunk 9]
```diff
@@ -237,6 +237,21 @@
     public static final Setting<TimeValue> INDEX_GC_DELETES_SETTING =
         Setting.timeSetting("index.gc_deletes", DEFAULT_GC_DELETES, new TimeValue(-1, TimeUnit.MILLISECONDS), Property.Dynamic,
             Property.IndexScope);
+
+    /**
+     * Specifies if the index should use soft-delete instead of hard-delete for update/delete operations.
+     */
+    public static final Setting<Boolean> INDEX_SOFT_DELETES_SETTING =
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/IndexSettings.java]: I tried but was unable to get it done with the current Validator. I think we need to extend the Validator to pass an entire Settings instance to make this possible (previous discussion https://github.com/elastic/elasticsearch/pull/25560#discussion_r125772288). Would it okay if I make this in a follow-up?

File: server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 10]
```diff
@@ -192,7 +193,25 @@ private synchronized boolean assertNoSeqNumberConflict(long seqNo, BytesReferenc
                         new BufferedChecksumStreamInput(data.streamInput(), "assertion"));
                 Translog.Operation prvOp = Translog.readOperation(
                         new BufferedChecksumStreamInput(previous.v1().streamInput(), "assertion"));
-                if (newOp.equals(prvOp) == false) {
+                // TODO: We haven't had timestamp for Index operations in Lucene yet, we need to loosen this check without timestamp.
+                // We don't store versionType in Lucene index, we need to exclude it from this check
+                final boolean sameOp;
+                if (newOp instanceof Translog.Index && prvOp instanceof Translog.Index) {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: I removed versionType from the comment. I will move this comparison to the Operation's equals.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 11]
```diff
@@ -451,19 +480,35 @@ public void syncTranslog() throws IOException {
...
+     */
     @Override
-    public int estimateTranslogOperationsFromMinSeq(long minSeqNo) {
-        return getTranslog().estimateTotalOperationsFromMinSeq(minSeqNo);
+    public int estimateNumberOfHistoryOperations(String source, MapperService mapperService, long startingSeqNo) throws IOException {
+        if (engineConfig.getIndexSettings().isSoftDeleteEnabled()) {
+            try (Translog.Snapshot snapshot =
+                     newLuceneChangesSnapshot(source, mapperService, Math.max(0, startingSeqNo), Long.MAX_VALUE, false)) {
+                return snapshot.totalOperations();
+            } catch (IOException ex) {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I removed this catch and added a tragic check when we create a new snapshot.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: There is a bug with the current implementation. I added a test and fixed this in 
https://github.com/elastic/elasticsearch/pull/33222/commits/4ed79074752f0abc82708e32c34fc0217be8238f

**Comment by dnhatn** [comment]
@s1monw and @jpountz I've addressed all your comments except the settings suggestion. Can you please have another look? Thank you!

File: server/src/main/java/org/elasticsearch/common/lucene/Lucene.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (s1monw) [reply server/src/main/java/org/elasticsearch/common/lucene/Lucene.java]: what is not clean on your side?

**Review by s1monw — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
Hmm. CI failed due to https://github.com/elastic/elasticsearch/issues/32299.

@elasticmachine retest this please.

**Comment by dnhatn** [comment]
Thanks so much @s1monw and @jpountz!

**Comment by dnhatn** [comment]
@elasticmachine run sample packaging tests please

**Comment by dnhatn** [comment]
@elasticmachine retest this please.

**Comment by dnhatn** [comment]
Another watcher test failure `./gradlew :x-pack:plugin:integTestRunner -Dtests.seed=3DA1EDA8F6DFAADB -Dtests.class=org.elasticsearch.xpack.test.rest.XPackRestIT -Dtests.method="test {p0=watcher/usage/10_basic/Test watcher usage stats output}" -Dtests.security.manager=true -Dtests.locale=nl -Dtests.timezone=Africa/Gaborone -Dcompiler.java=10 -Druntime.java=8 -Dtests.rest.blacklist=getting_started/10_monitor_cluster_health/*`

@elasticmachine test this please.

**Pull request merged**